### ISSUE TITLE
[Auto-upload Published] Show Publish button in Post List

### ISSUE THREAD

**Issue by shiki** [wordpress-mobile/WordPress-iOS#12321]
📣 Part of #12240. To decrease the risk of conflicts, this should use the #12317 PR branch as the base branch. Ideally, that branch should be merged first. Any fixes should target the `issue/12240-master-branch` branch. 

In #12317, we added the _Cancel_ button in the Post List. When the user presses the _Cancel_ button, the app will not try to automatically upload the post anymore. The app will then hide the _Cancel_ button and only show _Edit_ and _More_. 

<img src="https://user-images.githubusercontent.com/198826/63058517-23737e00-beaa-11e9-9a29-b055b427d101.png" width="320">

There used to be a _Retry_ button in the middle but I removed it since it seems to be confusing if it is shown after cancelation. 

In WPAndroid, we show a _Publish_ button after canceling.

![image](https://user-images.githubusercontent.com/198826/63058621-67ff1980-beaa-11e9-97d8-702b9253c2b0.png)

## Proposal

![image](https://user-images.githubusercontent.com/198826/63058974-66822100-beab-11e9-9754-01e54fccd552.png)

## Questions

- [x] The _Publish_ button is part of the spec described in #12240. Should we add this as part of the scope for iOS? Or will the current experience be good enough? 

@osullivanchris Hoping to get your input here. 🙂 

## Decisions 

- 2019-Aug-15 We'll include this in the scope as decided in https://github.com/wordpress-mobile/WordPress-iOS/issues/12321#issuecomment-521570525. 
- 2019-Aug-22 We had a chat about https://github.com/wordpress-mobile/WordPress-iOS/issues/12321#issuecomment-523392568 and we decided that we will be implementing **Option B** described in https://github.com/wordpress-mobile/WordPress-iOS/issues/12321#issuecomment-523097247. 
- 2019-Aug-26 We had some confusion around where to show the Publish button. We decided on having only a maximum of 3 buttons on the Post List. If there are more, the rest, preferably the _Publish_ and _Trash_, should be hidden under _More_. Please see [this illustration](https://user-images.githubusercontent.com/198826/63712173-5e56a980-c7fa-11e9-858e-0f44847d07ff.png).

## Discovered Tasks 

- [x] ~Pressing on the Publish button should mark the post as “confirmed for auto-uploading“. If offline, the post will be automatically uploaded later.~ #12382

**Comment by osullivanchris** [comment]
@shiki  the proposal you shared with the 'publish' button returning after cancelling looks correct to me and what I would expect. So I'm happy to go with that!

**Comment by shiki** [comment]
@osullivanchris I was talking with @yaelirub about this yesterday. It looks like I failed to communicate what I meant for “showing the Publish button“. I meant to communicate that we will follow how Android's Publish button works. Android has been showing the Publish button even before the Cancel button even existed. 

In Android, the _Publish_ button is shown when any of these conditions are true ([ref](https://github.com/wordpress-mobile/WordPress-Android/blob/master-auto-upload/WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt#L327)):

- The post is a draft. 
- The post was previously published and has local changes but the user **canceled** auto-uploading
- The upload failed and the user cannot _Cancel_ it anymore. This happens when we reached the maximum number of retries. 

However, Yael understood this ticket as “show the Publish button after canceling only“, so clearly I failed in communicating this well. 

There's a significant difference in the implementation of both. For example, the former does not require any change in the database while the latter requires a new property (flag) in the database. 

I'd like to hear what you think. Which one makes sense? Should we follow Android or implement it differently on iOS? 😐

**Comment by osullivanchris** [comment]
> However, Yael understood this ticket as “show the Publish button after canceling only“, so clearly I failed in communicating this well.

That's what I thought it meant :) 

> The upload failed and the user cannot Cancel it anymore. This happens when we reached the maximum number of retries.

That's fine. The error message will also have changed in this scenario right? So it won't be saying its going to publish anymore.

> There's a significant difference in the implementation of both. For example, the former does not require any change in the database while the latter requires a new property (flag) in the database.

I don't really understand this so I guess its a technical decision?

> I'd like to hear what you think. Which one makes sense? Should we follow Android or implement it differently on iOS? 😐

Is there any alternative to the proposed solution? I can't think of what other way we would implement it :)

**Comment by shiki** [comment]
> That's what I thought it meant :)

Okay. I suppose we could try and do that then. 

> Is there any alternative to the proposed solution? I can't think of what other way we would implement it :)

The other way is **not doing anything** at all. 🙂 To reconfirm auto-upload, our users would need to:

1. Edit the post 
2. Click on the Publish button again 

The _Post List_ would look empty though: 

<img src="https://user-images.githubusercontent.com/198826/63058517-23737e00-beaa-11e9-9a29-b055b427d101.png" width="320">

**Comment by shiki** [comment]
> Is there any alternative to the proposed solution?

Oh, sorry. To clarify, there are three options:

a. Show Publish only when Cancel was pressed.
b. Show Publish like Android: whenever the conditions match as described in https://github.com/wordpress-mobile/WordPress-iOS/issues/12321#issuecomment-523022814.
c. Not showing the Publish button at all. 

But I believe this discussion is leading towards **a**. Please let me know otherwise. 🙂

**Comment by yaelirub** [comment]
I wonder if C makes more sense. Not to say that we won't do it but to delay it after Post Logic Improvements (I haven't seen the scope for that yet).

With that said, I also realize that we want to align features on both platforms so I'd go with A.

Leaving the decision to you, @shiki .

**Comment by osullivanchris** [comment]
If I understand correctly, if we go with option A or C there will be some scenarios where the user is shown an error message and no option to publish the post. That feels like we're creating a dead end for the user. I don't really see what the advantage of that option is. 

So (b) to me sounds like the only viable UX. But I might have misunderstood something, or maybe the cost is huge with (b). @shiki if you want to jump on a quick call to talk this through today it might be quicker to get to the bottom of.

**Comment by shiki** [comment]
Done in #12375.
